### PR TITLE
CheckStyle-Rule-Extension: 6, 7, 8

### PR DESCRIPTION
**Pull request by mssfang**
- Check 6:  **@ServiceClientBuilder**
  - All service client builder should be named <ServiceName>ClientBuilder and annotated with @ServiceClientBuilder.
  - Has a method 'buildClient()' to build a synchronous client,
  - Has a method 'buildAsyncClient()' to build an asynchronous client
- Check 7:  **@ServiceInterface**
  - The annotation property 'name' should be non-empty
  - The length of value of property 'name' should be less than 20 characters and without space
- Check 8: **No OpenCensus dependency**
  - OpenCensus should only be depended on from within the tracing-opencensus module. It should not be used in any other class throughout the Java SDK.

### CODE REVIEW

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                if (hasServiceClientBuilderAnnotation) {
+                    // Checks if the ANNOTATION has property named 'serviceClients'
+                    if (!hasServiceClientsAnnotationProperty(serviceClientAnnotationToken)) {
+                        log(serviceClientAnnotationToken, String.format(
+                            "Annotation @%s should have ''serviceClients'' as property of annotation and should list all of the service clients it can build.",
+                            SERVICE_CLIENT_BUILDER));
+                    }
+                    // HAS @ServiceClientBuilder annotation but NOT named the class <ServiceName>ClientBuilder
+                    if (!className.endsWith("ClientBuilder")) {
+                        log(token, String.format("@ServiceClientBuilder class ''%s''should be named <ServiceName>ClientBuilder.", className));
```
Comment (JonathanGiles) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]:
```suggestion
                        log(token, String.format("Class annotated with @ServiceClientBuilder ''%s''should be named <ServiceName>ClientBuilder.", className));
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                final DetailAST serviceClientAnnotationToken = getServiceClientAnnotation(token);
+                final String className = token.findFirstToken(TokenTypes.IDENT).getText();
+
+                hasServiceClientBuilderAnnotation = serviceClientAnnotationToken != null;
+
+                if (hasServiceClientBuilderAnnotation) {
+                    // Checks if the ANNOTATION has property named 'serviceClients'
+                    if (!hasServiceClientsAnnotationProperty(serviceClientAnnotationToken)) {
+                        log(serviceClientAnnotationToken, String.format(
+                            "Annotation @%s should have ''serviceClients'' as property of annotation and should list all of the service clients it can build.",
```
Comment (JonathanGiles) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: You don't need to use %s for ServiceClientBuilder - it's a fixed string.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        if (!hasServiceClientBuilderAnnotation) {
+            return;
+        }
+
+        if (!hasAsyncClientBuilder) {
+            log(root, String.format("Every Missing an asynchronous method, ''%s''", BUILD_ASYNC_CLIENT));
+        }
+        if (!hasClientBuilder) {
+            log(root, String.format("Missing a synchronous method, ''%s''", BUILD_CLIENT));
+        }
```
Comment (JonathanGiles) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: These two strings need improving - they are hard to read.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java (hunk 2 above)
Comment (mssfang) [reply eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: My bad. This log will be removed because of 'serviceClients' is required when they add @ServiceClientBuilder

**Comment by mssfang** [comment]
Take off rule 5 check and prioritize on rule 8 and 9

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ *  1) All service client builder should be named <ServiceName>ClientBuilder and annotated with @ServiceClientBuilder.
+ *  2) Has a method 'buildClient()' to build a synchronous client,
+ *  3) Has a method 'buildAsyncClient()' to build an asynchronous client
+ */
+public class ServiceClientBuilderCheck extends AbstractCheck {
+    private static final String SERVICE_CLIENT_BUILDER = "ServiceClientBuilder";
+    private static final String BUILD_CLIENT = "buildClient";
+    private static final String BUILD_ASYNC_CLIENT = "buildAsyncClient";
+
+    private static boolean hasServiceClientBuilderAnnotation;
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Make these non-static

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        }
+    }
+
+    /**
+     * Checks if the class is annotated with @ServiceClientBuilder.
+     *
+     * @param classDefToken the CLASS_DEF AST node
+     * @return the annotation node if the class is annotated with @ServiceClientBuilder, null otherwise.
+     */
+    private DetailAST getServiceClientAnnotation(DetailAST classDefToken) {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Change name to `getServiceClientBuilderAnnotation`

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            TokenTypes.CLASS_DEF,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
+                final DetailAST serviceClientAnnotationToken = getServiceClientAnnotation(token);
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]:
```suggestion
                final DetailAST serviceClientBuilderAnnotationToken = getServiceClientBuilderAnnotation(token);
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                    }
+                } else {
+                    // No @ServiceClientBuilder annotation but HAS named the class <ServiceName>ClientBuilder
+                    if (className.endsWith("ClientBuilder")) {
+                        log(token, String.format("Class ''%s'' should be annotated with @ServiceClientBuilder.", className));
+                    }
+                }
+                break;
+            case TokenTypes.METHOD_DEF:
+                if (!hasServiceClientBuilderAnnotation) {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Is the order of `visitToken` call as per the array returned by `getRequiredTokens()`? If not, visitToken might be called with `METHOD_DEF` first and then `CLASS_DEF`. If this happens, `hasServiceClientBuilderAnnotation` would not be updated yet and will be `false` resulting in skipping method validation.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                    if (!"ServiceInterface".equals(annotationChild.getText())) {
+                        break;
+                    } else {
+                        serviceInterfaceAnnotationNode = ast;
+                    }
+                }
+
+                // ANNOTATION_MEMBER_VALUE_PAIR
+                if (annotationChild.getType() == TokenTypes.ANNOTATION_MEMBER_VALUE_PAIR) {
+                    if ("name".equals(annotationChild.findFirstToken(TokenTypes.IDENT).getText())) {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java]: You can merge these two `if` conditions into one

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                        nameValue = getNamePropertyValue(annotationChild.findFirstToken(TokenTypes.EXPR));
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java]: You can `break` after you have found the `nameValue`

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+        // Checks the rules:
+        // Skip the check if no @ServiceInterface annotation found
+        if (serviceInterfaceAnnotationNode == null) {
+            return;
+        }
+
+        // 'name' is required at @ServiceInterface
+        // 'name' should not be empty
+        if (nameValue.isEmpty()) {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java]: I think you can use a simple regex to check all these 3 rules at once.

```
Pattern serviceNamePattern = Pattern.compile("^[a-zA-Z0-9]{1,10}$");
if (!serviceNamePattern.matcher(nameValue).find()) {
	log(serviceInterfaceAnnotationNode, String.format(
		"The ''name'' property of @ServiceInterface, ''%s'' should be non-empty, alphanumeric and not more than 10 characters",
		nameValue));
}
```

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+     *
+     * @param interfaceDefToken INTERFACE_DEF AST node
+     */
+    private void checkServiceInterface(DetailAST interfaceDefToken) {
+        DetailAST serviceInterfaceAnnotationNode = null;
+        String nameValue = null;
+
+        DetailAST modifiersToken = interfaceDefToken.findFirstToken(TokenTypes.MODIFIERS);
+        // Find the @ServiceInterface and the property 'name' and the corresponding value
+        for (DetailAST ast = modifiersToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java]: Consider extracting some parts of this for loop into smaller methods for better readability.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java (hunk 7 above)
Comment (mssfang) [reply eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Usually, we will put the inner nested class at the end of class def.  But it could still have a scenario that an inner nested class before methods in the same class. I modified the code to support the scenario in the new commit soon.

**Comment by mssfang** [comment]
There are 116 checkstyle errors produced by rule 9. Fixing those now

**Comment by mssfang** [comment]
No 9 is still experiencing the IO exception failed which is the same problem about reflection. But the existing rule implementation is ready for review.

File: core/azure-core/src/main/java/com/azure/core/implementation/RestProxy.java [hunk 12]
```diff
@@ -497,8 +501,9 @@ protected Object handleHttpResponse(final HttpRequest httpRequest, Mono<HttpDeco
     }
 
     protected Object handleResumeOperation(HttpRequest httpRequest, OperationDescription operationDescription, SwaggerMethodParser methodParser, Type returnType, Context context)
-        throws Exception {
-        throw new Exception("The resume operation is not available in the base RestProxy class.");
+        throws RuntimeException {
```
Comment (JonathanGiles) [inline core/azure-core/src/main/java/com/azure/core/implementation/RestProxy.java]: Don't declare throwing a runtime exception

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+import java.util.Stack;
+
+/**
+ * The @ServiceClientBuilder class should have the following rules:
+ *  1) All service client builder should be named <ServiceName>ClientBuilder and annotated with @ServiceClientBuilder.
+ *  2) Has a method 'buildClient()' to build a synchronous client,
+ *  3) Has a method 'buildAsyncClient()' to build an asynchronous client
```
Comment (JonathanGiles) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Not all builders need to have both methods. For example, Event Hubs is only async at the moment. There should be no other methods starting with 'build' other than these two.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ThrownClientLoggerCheck.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                        if (modifiersToken.branchContains(TokenTypes.LITERAL_STATIC)) {
+                            return;
+                        }
+                        break;
+                    }
+                    parentToken = parentToken.getParent();
+                }
+
+                // non static method and class
+                log(token, "To throws an exception, must do it through a ''logger.logAndThrow'', rather than "
```
Comment (JonathanGiles) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ThrownClientLoggerCheck.java]: s/throws/throw

**Comment by conniey** [comment]
I think these checks might have to be split up into another PR. (ie. Check 9 about logAndThrow will result in a PR that touches many files.)

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java (hunk 13 above)
Comment (mssfang) [reply eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Does every @ServiceClientBuilder class should have one of buildClient or buildAsyncClient method? or it could be none.

**Comment by mssfang** [comment]
Cleaned up and ready for review again. :>

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java (hunk 13 above)
Comment (JonathanGiles) [reply eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Probably the best answer is that there must be at least one method named build*Client(), where * could be any string zero or more characters long. That is, buildClient() is valid, so is buildAsyncClient(), buildBlobClient(), buildBlobAsyncClient(), etc. So, we should expect at least one of these.

Additionally, I think for now we could also have the rule enforce the requirement that any method that starts with `build` must also end with `Client`.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+                }
+                break;
+            case TokenTypes.METHOD_DEF:
+                if (!hasServiceClientBuilderAnnotation) {
+                    return;
+                }
+
+                final String methodName = token.findFirstToken(TokenTypes.IDENT).getText();
+                // method name has prefix 'build' but not 'buildClient' or 'buildAsyncClient'
+                if (methodName.startsWith("build") && !BUILD_ASYNC_CLIENT.equals(methodName) && !BUILD_CLIENT.equals(methodName)) {
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: There should also be a check to ensure there is at least one method that starts with "build".

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.INTERFACE_DEF
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.INTERFACE_DEF:
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceInterfaceCheck.java]: There is just one `case` here. We can replace this `switch` statement with `if`.

File: eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void leaveToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.CLASS_DEF:
```
Comment (srnagar) [inline eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientBuilderCheck.java]: Same here. Only one `case` here - replace `switch` with `if`.

**Comment by mssfang** [comment]
PR is ready for reviewing again

File: sdk/keyvault/azure-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/SecretService.java [hunk 18]
```diff
@@ -39,7 +39,7 @@
  * This is package-private so that these REST calls are transparent to the user.
  */
 @Host("{url}")
-@ServiceInterface(name = "KeyVaultSecrets")
+@ServiceInterface(name = "KVSecrets")
```
Comment (JonathanGiles) [inline sdk/keyvault/azure-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/SecretService.java]: I don't like this change. You should change the rule to make it zero to 20 characters, and revert this change.
